src/kernel.jl

```diff
@@ -84,14 +84,148 @@ function set_arg!(k::Kernel, idx::Integer, arg::LocalMem)
+inbuild_alignement{T}(x::T) = inbuild_alignement(T)
+
+function cl_alignement(x)
+    is_cl_inbuild(x) ? inbuild_alignement(x) : cl_sizeof(x)
+end
+
+function advance_aligned(offset, alignment)
+    (offset == 0 || alignment == 0) && return 0
+    if offset % alignment != 0
+        npad = ((div(offset, alignment) + 1) * alignment) - offset
+        offset += npad
+    end
+    offset
+end
+
+
+
+Base.@pure function _cl_sizeof{T}(::Type{T}, offset = 0)
+    align, size = if is_cl_inbuild(T) || nfields(T) == 0
+        align, size = inbuild_alignement(T), sizeof(T)
+        if offset == 0
+            return align
+        end
+        align, size
+    else
+        nextoffset = offset
+        for field in fieldnames(T)
+            xT = fieldtype(T, field)
+            nextoffset = _cl_sizeof(xT, nextoffset)
+        end
+        size = nextoffset - offset
+        size, size
+    end
+    offset = advance_aligned(offset, align)
+    offset += size
+    offset
+end
+
+Base.@generated function cl_sizeof{T}(x::T)
+    size = _cl_sizeof(T, 0)
+    :($size)
+end
+Base.@generated function cl_sizeof{T}(x::Type{T})
+    size = _cl_sizeof(T, 0)
+    :($size)
+end
+
+@generated function aligned_convert(x)
+    offset = Ref(0); elements = []; fields = []
+    _aligned_convert!(x, offset, elements, fields, :x)
+    ret = if length(elements) == 1 # no conversion happened
+        return :(x, $(offset[]))
+    else
+        tupl = Expr(:tuple)
+        tupl.args = first.(elements)
+        expr = quote
+            $(fields...) # hoisted field loads
+            $tupl, $(offset[])
+        end
+        expr
+    end
+end
+
+@generated function aligned_convert{T}(x::Type{T})
+    offset = Ref(0); elements = []; fields = []
+    _aligned_convert!(T, offset, elements, fields, :x)
+    ret = if length(elements) == 1 # no conversion happened
+        return :(T, $(offset[]))
+    else
+        expr = quote
+            Tuple{$(last.(elements)...)}, $(offset[])
+        end
+        expr
+    end
+end
+
+function _aligned_convert!(x, offset = Ref(0), elements = [], fields = [], fieldname = gensym(:field))
+    alignment = cl_alignement(x)
+    if alignment != 0 && offset[] % alignment != 0
+        npad = ((div(offset[], alignment) + 1) * alignment) - offset[]
+        pad = Pad{npad}()
+        offset[] += npad
+        push!(elements, :(Pad{$npad}()) => Pad{npad})
+    end
+    if !is_cl_inbuild(x) && nfields(x) > 0
+        for field in fieldnames(x)
+            current_field = gensym(string(field))
+            push!(fields, :($current_field = getfield($fieldname, $(QuoteNode(field)))))
+            xelem = fieldtype(x, field)
+            _aligned_convert!(xelem, offset, elements, fields, current_field)
+        end
+    else
+        push!(elements, fieldname => x)
+        offset[] += sizeof(x)
+    end
+    return
+end
+
+
+
 #TODO: type safe calling of set args for kernel (with clang)
+# is 1024 a good number?
+const _arg_tmp_buffer = Vector{Int8}(1024)
```

Comment:
I feel like we're far away from multi threaded kernel launches  ;) but if it happens we can still change it!